File: DFS_BFS/키패드누르기.cpp
#include <string>
#include <vector>

using namespace std;

string answer = "";

int calculateDist(int mid, int a)
{
    if (mid == a)
        return 0;
    if (mid == 2)
    {
        if (a == 1 || a == 3 || a == 5)
            return 1;
        if (a == 4 || a == 6 || a == 8)
            return 2;
        if (a == 7 || a == 9 || a == 0)
            return 3;
        if (a == 11 || a == 12)
            return 4;
    }
    else if (mid == 5)
    {
        if (a == 1 || a == 3)
            return 2;
        if (a == 4 || a == 6 || a == 2 || a == 8)
            return 1;
        if (a == 7 || a == 9 || a == 0)
            return 2;
        if (a == 11 || a == 12)
            return 3;
    }
    else if (mid == 8)
    {
        if (a == 1 || a == 3)
            return 3;
        if (a == 4 || a == 6 || a == 2)
            return 2;
        if (a == 7 || a == 9 || a == 5 || a == 0)
            return 1;
        if (a == 11 || a == 12)
            return 2;
    }
    else if (mid == 0)
    {
        if (a == 1 || a == 3)
            return 4;
        if (a == 4 || a == 6 || a == 2)
            return 3;
        if (a == 7 || a == 9 || a == 5)
            return 2;
        if (a == 11 || a == 12 || a == 8)
            return 1;
    }
}

void dfs(vector<int> numbers, string hand, int i, int pastLeft, int pastRight)
{
    if (i == numbers.size())
        return;

    if (numbers[i] == 1 || numbers[i] == 4 || numbers[i] == 7)
    {
        answer += "L";
        pastLeft = numbers[i];
    }
    else if (numbers[i] == 3 || numbers[i] == 6 || numbers[i] == 9)
    {
        answer += "R";
        pastRight = numbers[i];
    }
    else
    {
        int mid = numbers[i];
        int leftDist = calculateDist(mid, pastLeft);
        int rightDist = calculateDist(mid, pastRight);
        if (leftDist > rightDist)
        {
            answer += "R";
            pastRight = mid;
        }
        else if (leftDist < rightDist)
        {
            answer += "L";
            pastLeft = mid;
        }
        else
        {
            if (hand == "left")
            {
                answer += "L";
                pastLeft = mid;
            }
            else
            {
                answer += "R";
                pastRight = mid;
            }
        }
    }
    dfs(numbers, hand, i + 1, pastLeft, pastRight);
}

string solution(vector<int> numbers, string hand)
{
    dfs(numbers, hand, 0, 11, 12);
    return answer;
}

File: String/키패드누르기.js
function get_dist(hand_index, number_index, flag) {
    // hand가 가운데에 있으면
    if (flag) {
        return Math.abs(hand_index - number_index);
    } 
    // hand가 양 옆에 있으면
    else {
        return Math.abs(hand_index - number_index) + 1;
    }
}

function solution(numbers, hand) {
    numbers = numbers.map(number => number === 0 ? 11 : number);
    
    const left_key = [1, 4, 7, 10];
    const center_key = [2, 5, 8, 11];
    const right_key = [3, 6, 9, 12];
    
    // Start hand loc
    let left_hand = 10;
    let right_hand = 12;
    
    // answer
    let answer = "";
    for (const number of numbers){
        const left_index = left_key.indexOf(number);
        const right_index = right_key.indexOf(number);
        const center_index = center_key.indexOf(number);
        
        // to Left
        if (left_index !== -1) {
            answer += "L";
            left_hand = number;
        }
        // to Right
        else if (right_index !== -1) {
            answer += "R";
            right_hand = number;
        }
        // to Center
        else {
            const left_flag = center_key.indexOf(left_hand) === -1 ? false : true;
            const right_flag = center_key.indexOf(right_hand) === -1 ? false : true;
            
            const left_hand_index = left_flag ? center_key.indexOf(left_hand) : left_key.indexOf(left_hand);
            const right_hand_index = right_flag ? center_key.indexOf(right_hand) : right_key.indexOf(right_hand);
            
            const left_dist = get_dist(left_hand_index, center_index, left_flag);
            const right_dist = get_dist(right_hand_index, center_index, right_flag);
            
            // left_dist < right_dist : select left_hand
            if (left_dist < right_dist) {
                answer += "L";
                left_hand = number;
            } 
            // left_dist > right_dist : select right_hand
            else if (left_dist > right_dist) {
                answer += "R";
                right_hand = number;
            }
            // left_dist === right_dist : select main hand
            else {
                if (hand === "left") {
                    answer += "L";
                    left_hand = number;
                } else {
                    answer += "R";
                    right_hand = number;
                }
            }
        }
    }
    return answer;
}